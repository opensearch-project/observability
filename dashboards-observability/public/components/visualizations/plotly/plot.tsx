/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import plotComponentFactory from 'react-plotly.js/factory';
import Plotly from 'plotly.js-dist';
import { uiSettingsService } from '../../../../common/utils';

interface PltProps {
  data: Plotly.Data[];
  layout?: Partial<Plotly.Layout>;
  config?: Partial<Plotly.Config>;
  onHoverHandler?: (event: Readonly<Plotly.PlotMouseEvent>) => void;
  onUnhoverHandler?: (event: Readonly<Plotly.PlotMouseEvent>) => void;
  onClickHandler?: (event: Readonly<Plotly.PlotMouseEvent>) => void;
  height?: string;
  dispatch?: (props: any) => void;
}

export function Plt(props: PltProps) {
  const PlotComponent = plotComponentFactory(Plotly);
  const darkLayout = uiSettingsService.get('theme:darkMode')
    ? {
        paper_bgcolor: '#1D1E24',
        plot_bgcolor: '#1D1E24',
        font: {
          color: '#DFE5EF',
        },
      }
    : {};

  const finalLayout = {
    autosize: true,
    margin: {
      l: 30,
      r: 5,
      b: 30,
      t: 5,
      pad: 4,
    },
    barmode: 'stack',
    legend: {
      orientation: 'h',
      traceorder: 'normal',
    },
    showlegend: false,
    hovermode: 'closest',
    xaxis: {
      showgrid: true,
      zeroline: false,
      rangemode: 'normal',
      automargin: true,
    },
    yaxis: {
      showgrid: true,
      zeroline: false,
      rangemode: 'normal',
    },
    ...darkLayout,
    ...props.layout,
  };

  const finalConfig = {
    displayModeBar: false,
    ...props.config,
  };

  return (
    <PlotComponent
      divId="explorerPlotComponent"
      data={props.data}
      style={{ width: '100%', height: props.height || '100%' }}
      onHover={props.onHoverHandler}
      onUnhover={props.onUnhoverHandler}
      onClick={props.onClickHandler}
      useResizeHandler
      config={finalConfig}
      layout={finalLayout}
    />
  );
}
