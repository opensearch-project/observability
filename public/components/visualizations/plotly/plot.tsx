/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import plotComponentFactory from 'react-plotly.js/factory';
import Plotly from 'plotly.js-dist';

interface PltProps {
  data: Plotly.Data[];
  layout?: Partial<Plotly.Layout>;
  config?: Partial<Plotly.Config>;
  onHoverHandler?: (event: Readonly<Plotly.PlotMouseEvent>) => void;
  onUnhoverHandler?: (event: Readonly<Plotly.PlotMouseEvent>) => void;
  height?: string;
}

export function Plt(props: PltProps) {
  const PlotComponent = plotComponentFactory(Plotly);
  return (
    <PlotComponent
      data={props.data}
      style={{ width: '100%', height: props.height || '100%' }}
      onHover={props.onHoverHandler}
      onUnhover={props.onUnhoverHandler}
      useResizeHandler
      config={{ 
        displayModeBar: false,
        ...props.config
      }}
      layout={{
        autosize: true,
        margin: {
          l: 30,
          r: 5,
          b: 30,
          t: 5,
          pad: 4,
        },
        barmode: 'stack',
        legend: {
          orientation: 'h',
          traceorder: 'normal',
        },
        showlegend: false,
        hovermode: 'closest',
        xaxis: {
          showgrid: true,
          zeroline: false,
          rangemode: 'normal',
          automargin: true
        },
        yaxis: {
          showgrid: true,
          zeroline: false,
          rangemode: 'normal'
        },
        ...props.layout,
      }}
    />
  );
}
