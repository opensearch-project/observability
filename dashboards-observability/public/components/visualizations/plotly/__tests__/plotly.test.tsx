/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { Plt } from '../plot';
import { LAYOUT_CONFIG } from '../../../../../test/event_analytics_constants';

describe('Ploty base component', () => {
  configure({ adapter: new Adapter() });

  it('Renders Ploty base component', async () => {
    
    const barValues = [{
      displaylogo: false,
      marker: {
        color: [
          "#3CA1C7",
          "#8C55A3",
          "#DB748A",
          "#F2BE4B"
        ]
      },
      name: "avg(FlightDelayMin)",
      responsive: true,
      type: 'bar',
      x: [
        "JetBeats",
        "Logstash Airways",
        "OpenSearch Dashboards Airlines",
        "OpenSearch-Air"
      ],
      y: [
        45.957544288332315,
        49.55268688081657,
        46.368274582560296,
        47.41304347826087
      ]
    }];

    const wrapper = mount(
      <Plt
        data={barValues}
        layout={{
          xaxis: {
            showgrid: false,
            visible: true,
          },
          yaxis: {
            showgrid: false,
            visible: true,
          },
          ...LAYOUT_CONFIG,
        }}
        config={{}}
      />
    );
    
    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});