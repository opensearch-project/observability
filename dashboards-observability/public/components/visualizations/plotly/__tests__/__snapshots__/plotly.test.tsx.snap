// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ploty base component Renders Ploty base component 1`] = `
<Plt
  config={Object {}}
  data={
    Array [
      Object {
        "displaylogo": false,
        "marker": Object {
          "color": Array [
            "#3CA1C7",
            "#8C55A3",
            "#DB748A",
            "#F2BE4B",
          ],
        },
        "name": "avg(FlightDelayMin)",
        "responsive": true,
        "type": "bar",
        "x": Array [
          "JetBeats",
          "Logstash Airways",
          "OpenSearch Dashboards Airlines",
          "OpenSearch-Air",
        ],
        "y": Array [
          45.957544288332315,
          49.55268688081657,
          46.368274582560296,
          47.41304347826087,
        ],
      },
    ]
  }
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
      "xaxis": Object {
        "showgrid": false,
        "visible": true,
      },
      "yaxis": Object {
        "showgrid": false,
        "visible": true,
      },
    }
  }
>
  <PlotlyComponent
    config={
      Object {
        "displayModeBar": false,
      }
    }
    data={
      Array [
        Object {
          "displaylogo": false,
          "marker": Object {
            "color": Array [
              "#3CA1C7",
              "#8C55A3",
              "#DB748A",
              "#F2BE4B",
            ],
          },
          "name": "avg(FlightDelayMin)",
          "responsive": true,
          "type": "bar",
          "x": Array [
            "JetBeats",
            "Logstash Airways",
            "OpenSearch Dashboards Airlines",
            "OpenSearch-Air",
          ],
          "y": Array [
            45.957544288332315,
            49.55268688081657,
            46.368274582560296,
            47.41304347826087,
          ],
        },
      ]
    }
    debug={false}
    divId="explorerPlotComponent"
    layout={
      Object {
        "autosize": true,
        "barmode": "stack",
        "colorway": Array [
          "#8C55A3",
        ],
        "height": 220,
        "hovermode": "closest",
        "legend": Object {
          "orientation": "h",
          "traceorder": "normal",
        },
        "margin": Object {
          "b": 15,
          "l": 60,
          "pad": 0,
          "r": 10,
          "t": 30,
        },
        "showlegend": true,
        "xaxis": Object {
          "showgrid": false,
          "visible": true,
        },
        "yaxis": Object {
          "showgrid": false,
          "visible": true,
        },
      }
    }
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
    useResizeHandler={true}
  >
    <div
      id="explorerPlotComponent"
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </PlotlyComponent>
</Plt>
`;
